File: src/api/file.js
/**
 * 文件管理相关 API
 */

import { invoke } from '@tauri-apps/api/tauri'

/**
 * 获取远程目录文件列表
 * @param {Object} params - 参数
 * @param {string} params.serverId - 服务器ID
 * @param {string} params.path - 远程路径，如 "/home/user"
 * @returns {Promise<{files: Array<{name: string, type: 'file'|'directory', size: number, modified: string, path: string}>}>}
 */
export async function listRemoteDirectory(params) {
  try {
    const result = await invoke('list_remote_directory', {
      serverId: params.serverId,
      path: params.path
    })
    return result
  } catch (error) {
    console.error('获取远程目录列表失败:', error)
    throw new Error(error.message || '获取文件列表失败')
  }
}

/**
 * 上传文件到远程服务器
 * @param {Object} params - 参数
 * @param {string} params.serverId - 服务器ID
 * @param {string} params.localPath - 本地文件路径
 * @param {string} params.remotePath - 远程保存路径（目录）
 * @returns {Promise<{success: boolean}>}
 */
export async function uploadFile(params) {
  try {
    const result = await invoke('upload_file', {
      serverId: params.serverId,
      localPath: params.localPath,
      remotePath: params.remotePath
    })
    return result
  } catch (error) {
    console.error('上传文件失败:', error)
    throw new Error(error.message || '上传文件失败')
  }
}

/**
 * 批量上传文件到远程服务器
 * @param {Object} params - 参数
 * @param {string} params.serverId - 服务器ID
 * @param {string[]} params.localPaths - 本地文件路径数组
 * @param {string} params.remotePath - 远程保存路径（目录）
 * @returns {Promise<{success: boolean, count: number}>}
 */
export async function uploadFiles(params) {
  try {
    let successCount = 0
    let failCount = 0
    
    for (const localPath of params.localPaths) {
      try {
        await uploadFile({
          serverId: params.serverId,
          localPath: localPath,
          remotePath: params.remotePath
        })
        successCount++
      } catch (error) {
        console.error(`上传文件失败 ${localPath}:`, error)
        failCount++
      }
    }
    
    return {
      success: failCount === 0,
      count: successCount,
      failCount: failCount
    }
  } catch (error) {
    console.error('批量上传文件失败:', error)
    throw new Error(error.message || '批量上传文件失败')
  }
}

/**
 * 从远程服务器下载文件
 * @param {Object} params - 参数
 * @param {string} params.serverId - 服务器ID
 * @param {string} params.remotePath - 远程文件路径
 * @param {string} params.localPath - 本地保存路径（完整路径，包含文件名）
 * @returns {Promise<{success: boolean}>}
 */
export async function downloadFile(params) {
  try {
    const result = await invoke('download_file', {
      serverId: params.serverId,
      remotePath: params.remotePath,
      localPath: params.localPath
    })
    return result
  } catch (error) {
    console.error('下载文件失败:', error)
    throw new Error(error.message || '下载文件失败')
  }
}

/**
 * 创建远程目录
 * @param {Object} params - 参数
 * @param {string} params.serverId - 服务器ID
 * @param {string} params.path - 要创建的目录完整路径
 * @returns {Promise<{success: boolean}>}
 */
export async function createDirectory(params) {
  try {
    const result = await invoke('create_directory', {
      serverId: params.serverId,
      path: params.path
    })
    return result
  } catch (error) {
    console.error('创建目录失败:', error)
    throw new Error(error.message || '创建目录失败')
  }
}

/**
 * 删除远程文件/目录
 * @param {Object} params - 参数
 * @param {string} params.serverId - 服务器ID
 * @param {string[]} params.paths - 要删除的文件/目录路径数组
 * @returns {Promise<{success: boolean}>}
 */
export async function deleteFiles(params) {
  try {
    const result = await invoke('delete_files', {
      serverId: params.serverId,
      paths: params.paths
    })
    return result
  } catch (error) {
    console.error('删除文件失败:', error)
    throw new Error(error.message || '删除文件失败')
  }
}

/**
 * 重命名远程文件/目录
 * @param {Object} params - 参数
 * @param {string} params.serverId - 服务器ID
 * @param {string} params.oldPath - 原路径
 * @param {string} params.newPath - 新路径
 * @returns {Promise<{success: boolean}>}
 */
export async function renameFile(params) {
  try {
    const result = await invoke('rename_file', {
      serverId: params.serverId,
      oldPath: params.oldPath,
      newPath: params.newPath
    })
    return result
  } catch (error) {
    console.error('重命名文件失败:', error)
    throw new Error(error.message || '重命名文件失败')
  }
}

/**
 * 修改文件权限
 * @param {Object} params - 参数
 * @param {string} params.serverId - 服务器ID
 * @param {string} params.path - 文件路径
 * @param {string} params.mode - 权限模式（如 "755", "644"）
 * @returns {Promise<{success: boolean}>}
 */
export async function changeFileMode(params) {
  try {
    const result = await invoke('change_file_mode', {
      serverId: params.serverId,
      path: params.path,
      mode: params.mode
    })
    return result
  } catch (error) {
    console.error('修改文件权限失败:', error)
    throw new Error(error.message || '修改文件权限失败')
  }
}

File: src-tauri/src/file.rs
/**
 * 文件管理相关命令处理
 */

use serde::{Deserialize, Serialize};

/// 文件信息
#[derive(Debug, Serialize)]
pub struct FileInfo {
    pub name: String,
    #[serde(rename = "type")]
    pub file_type: String, // "file" | "directory"
    pub size: u64,
    pub modified: String, // ISO 8601 格式
    pub path: String,
}

/// 获取远程目录文件列表参数
#[derive(Debug, Deserialize)]
pub struct ListRemoteDirectoryParams {
    pub server_id: String,
    pub path: String,
}

/// 获取远程目录文件列表返回
#[derive(Debug, Serialize)]
pub struct ListRemoteDirectoryResult {
    pub files: Vec<FileInfo>,
}

/// 上传文件参数
#[derive(Debug, Deserialize)]
pub struct UploadFileParams {
    pub server_id: String,
    pub local_path: String,
    pub remote_path: String,
}

/// 上传文件返回
#[derive(Debug, Serialize)]
pub struct UploadFileResult {
    pub success: bool,
    pub message: Option<String>,
}

/// 下载文件参数
#[derive(Debug, Deserialize)]
pub struct DownloadFileParams {
    pub server_id: String,
    pub remote_path: String,
    pub local_path: String,
}

/// 下载文件返回
#[derive(Debug, Serialize)]
pub struct DownloadFileResult {
    pub success: bool,
    pub message: Option<String>,
}

/// 创建目录参数
#[derive(Debug, Deserialize)]
pub struct CreateDirectoryParams {
    pub server_id: String,
    pub path: String,
}

/// 创建目录返回
#[derive(Debug, Serialize)]
pub struct CreateDirectoryResult {
    pub success: bool,
    pub message: Option<String>,
}

/// 删除文件参数
#[derive(Debug, Deserialize)]
pub struct DeleteFilesParams {
    pub server_id: String,
    pub paths: Vec<String>,
}

/// 删除文件返回
#[derive(Debug, Serialize)]
pub struct DeleteFilesResult {
    pub success: bool,
    pub message: Option<String>,
}

/// 重命名文件参数
#[derive(Debug, Deserialize)]
pub struct RenameFileParams {
    pub server_id: String,
    pub old_path: String,
    pub new_path: String,
}

/// 重命名文件返回
#[derive(Debug, Serialize)]
pub struct RenameFileResult {
    pub success: bool,
    pub message: Option<String>,
}

/// 修改文件权限参数
#[derive(Debug, Deserialize)]
pub struct ChangeFileModeParams {
    pub server_id: String,
    pub path: String,
    pub mode: String,
}

/// 修改文件权限返回
#[derive(Debug, Serialize)]
pub struct ChangeFileModeResult {
    pub success: bool,
    pub message: Option<String>,
}

/// 获取远程目录文件列表
/// 
/// # 命令名称
/// `list_remote_directory`
/// 
/// # 参数
/// - `server_id`: 服务器ID
/// - `path`: 远程路径
/// 
/// # 返回
/// - `files`: 文件列表
#[tauri::command]
pub async fn list_remote_directory(params: ListRemoteDirectoryParams) -> Result<ListRemoteDirectoryResult, String> {
    // TODO: 实现实际的 SFTP 文件列表获取逻辑
    // 这里需要使用 ssh2 或类似库来获取文件列表
    
    // 模拟返回
    Ok(ListRemoteDirectoryResult {
        files: vec![
            FileInfo {
                name: ".".to_string(),
                file_type: "directory".to_string(),
                size: 0,
                modified: "2024-01-01T00:00:00Z".to_string(),
                path: params.path.clone(),
            },
            FileInfo {
                name: "..".to_string(),
                file_type: "directory".to_string(),
                size: 0,
                modified: "2024-01-01T00:00:00Z".to_string(),
                path: format!("{}/..", params.path),
            },
        ],
    })
}

/// 上传文件到远程服务器
/// 
/// # 命令名称
/// `upload_file`
/// 
/// # 参数
/// - `server_id`: 服务器ID
/// - `local_path`: 本地文件路径
/// - `remote_path`: 远程保存路径（目录）
/// 
/// # 返回
/// - `success`: 是否成功
/// - `message`: 消息（可选）
#[tauri::command]
pub async fn upload_file(params: UploadFileParams) -> Result<UploadFileResult, String> {
    // TODO: 实现实际的 SFTP 文件上传逻辑
    // 这里需要使用 ssh2 或类似库来上传文件
    
    Ok(UploadFileResult {
        success: true,
        message: Some("上传成功".to_string()),
    })
}

/// 从远程服务器下载文件
/// 
/// # 命令名称
/// `download_file`
/// 
/// # 参数
/// - `server_id`: 服务器ID
/// - `remote_path`: 远程文件路径
/// - `local_path`: 本地保存路径
/// 
/// # 返回
/// - `success`: 是否成功
/// - `message`: 消息（可选）
#[tauri::command]
pub async fn download_file(params: DownloadFileParams) -> Result<DownloadFileResult, String> {
    // TODO: 实现实际的 SFTP 文件下载逻辑
    // 这里需要使用 ssh2 或类似库来下载文件
    
    Ok(DownloadFileResult {
        success: true,
        message: Some("下载成功".to_string()),
    })
}

/// 创建远程目录
/// 
/// # 命令名称
/// `create_directory`
/// 
/// # 参数
/// - `server_id`: 服务器ID
/// - `path`: 要创建的目录完整路径
/// 
/// # 返回
/// - `success`: 是否成功
/// - `message`: 消息（可选）
#[tauri::command]
pub async fn create_directory(params: CreateDirectoryParams) -> Result<CreateDirectoryResult, String> {
    // TODO: 实现实际的 SFTP 目录创建逻辑
    
    Ok(CreateDirectoryResult {
        success: true,
        message: Some("创建目录成功".to_string()),
    })
}

/// 删除远程文件/目录
/// 
/// # 命令名称
/// `delete_files`
/// 
/// # 参数
/// - `server_id`: 服务器ID
/// - `paths`: 要删除的文件/目录路径数组
/// 
/// # 返回
/// - `success`: 是否成功
/// - `message`: 消息（可选）
#[tauri::command]
pub async fn delete_files(params: DeleteFilesParams) -> Result<DeleteFilesResult, String> {
    // TODO: 实现实际的 SFTP 文件删除逻辑
    
    Ok(DeleteFilesResult {
        success: true,
        message: Some("删除成功".to_string()),
    })
}

/// 重命名远程文件/目录
/// 
/// # 命令名称
/// `rename_file`
/// 
/// # 参数
/// - `server_id`: 服务器ID
/// - `old_path`: 原路径
/// - `new_path`: 新路径
/// 
/// # 返回
/// - `success`: 是否成功
/// - `message`: 消息（可选）
#[tauri::command]
pub async fn rename_file(params: RenameFileParams) -> Result<RenameFileResult, String> {
    // TODO: 实现实际的 SFTP 文件重命名逻辑
    
    Ok(RenameFileResult {
        success: true,
        message: Some("重命名成功".to_string()),
    })
}

/// 修改文件权限
/// 
/// # 命令名称
/// `change_file_mode`
/// 
/// # 参数
/// - `server_id`: 服务器ID
/// - `path`: 文件路径
/// - `mode`: 权限模式（如 "755", "644"）
/// 
/// # 返回
/// - `success`: 是否成功
/// - `message`: 消息（可选）
#[tauri::command]
pub async fn change_file_mode(params: ChangeFileModeParams) -> Result<ChangeFileModeResult, String> {
    // TODO: 实现实际的 chmod 逻辑
    
    Ok(ChangeFileModeResult {
        success: true,
        message: Some("修改权限成功".to_string()),
    })
}

